src/machine/machine_rp2040_spi.go

```diff
@@ -0,0 +1,326 @@
+// until all the bytes in the command packet have been received:
+//
+// 		spi.Tx(tx, nil)
+//
+// This form sends zeros, putting the result into the rx buffer. Good for reading a "result packet":
+//
+// 		spi.Tx(nil, rx)
+//
+// Remark: This implementation (RP2040) allows reading into buffer with a custom repeated
+// value on tx.
+//
+// 		spi.Tx([]byte{0xff}, rx) // may cause unwanted heap allocations.
+//
+// This form sends 0xff and puts the result into rx buffer. Useful for reading from SD cards
+// which require 0xff input on SI.
+func (spi SPI) Tx(w, r []byte) (err error) {
+	switch {
+	case w == nil:
+		// read only, so write zero and read a result.
+		err = spi.rx(r, 0)
+	case r == nil:
+		// write only
+		err = spi.tx(w)
+	case len(w) == 1 && len(r) > 1:
+		// Read with custom repeated value.
+		err = spi.rx(r, w[0])
+	default:
+		// write/read
+		err = spi.txrx(w, r)
+	}
+	return err
+}
+
+// Write a single byte and read a single byte from TX/RX FIFO.
+func (spi SPI) Transfer(w byte) (byte, error) {
+	for !spi.isWritable() {
+	}
+
+	spi.Bus.SSPDR.Set(uint32(w))
+
+	for !spi.isReadable() {
+	}
+	return uint8(spi.Bus.SSPDR.Get()), nil
+}
+
+func (spi SPI) SetBaudRate(br uint32) error {
+	const freqin uint32 = 125 * MHz
+	const maxBaud uint32 = 66.5 * MHz // max output frequency is 66.5MHz on rp2040. see Note page 527.
+	// Find smallest prescale value which puts output frequency in range of
+	// post-divide. Prescale is an even number from 2 to 254 inclusive.
+	var prescale, postdiv uint32
+	for prescale = 2; prescale < 255; prescale += 2 {
+		if freqin < (prescale+2)*256*br {
+			break
+		}
+	}
+	if prescale > 254 || br > maxBaud {
+		return errSPIBaud
+	}
+	// Find largest post-divide which makes output <= baudrate. Post-divide is
+	// an integer in the range 1 to 256 inclusive.
+	for postdiv = 256; postdiv > 1; postdiv-- {
+		if freqin/(prescale*(postdiv-1)) > br {
+			break
+		}
+	}
+	spi.Bus.SSPCPSR.Set(prescale)
+	spi.Bus.SSPCR0.ReplaceBits((postdiv-1)<<rp.SPI0_SSPCR0_SCR_Pos, rp.SPI0_SSPCR0_SCR_Msk, 0)
+	return nil
+}
+
+func (spi SPI) GetBaudRate() uint32 {
+	const freqin uint32 = 125 * MHz
+	prescale := spi.Bus.SSPCPSR.Get()
+	postdiv := ((spi.Bus.SSPCR0.Get() & rp.SPI0_SSPCR0_SCR_Msk) >> rp.SPI0_SSPCR0_SCR_Pos) + 1
+	return freqin / (prescale * postdiv)
+}
+
+// Configure is intended to setup/initialize the SPI interface.
+// Default baudrate of 115200 is used if Frequency == 0. Default
+// word length (data bits) is 8. SPI0 default bus is
+//  SCK = GPIO18
+//  SDO = GPIO19
+//  SDI = GPIO16
+// No pin configuration is needed of SCK, SDO and SDI needed after calling Configure.
+func (spi SPI) Configure(config SPIConfig) error {
+	const defaultBaud uint32 = 115200
+	if config.SCK == 0 {
+		switch spi.Bus {
+		case rp.SPI0:
+			// These pins appear darkened "default" on pico pinout pictures on the web.
+			config.SCK = GPIO18
```

Comment:
This will overwrite any values passed in by the programmer. This means the a board such as the RP2040 Nano Connect will not be able to use the onboard WiFiNina chip.

Reply:
These are predetermined "default" values in case a zero-valued `SPIConfig` struct is passed to `Configure`. I'm not against removing it but it's nice to have a shorthand initialization, I know I've spent too much time mixing up MISO and MOSI.

Reply:
Alright, I think I understand the importance of changing this now. The issue with defining pins like so:
```go
       SPI0_SCK_PIN Pin = GPIO18
       SPI0_SDO_PIN Pin = GPIO19
       SPI0_SDI_PIN Pin = GPIO16
```
is that these are not the only SPI0 pins available. Maybe we should call them `DEFAULT_SPI0_SCK` and such?

Reply:
Ideally those constants should be defined in the board file, and then used in the machine file. That way each board can specify which pins are assigned to which function. A bunch of the Adafruit SAMD21/51 boards are setup this way.